client/modules/IDE/pages/IDEView.jsx

```diff
@@ -66,296 +70,212 @@ function WarnIfUnsavedChanges() {
 
 export const CmControllerContext = React.createContext({});
 
-class IDEView extends React.Component {
-  constructor(props) {
-    super(props);
+const IDEView = (props) => {
+  const ide = useSelector((state) => state.ide);
+  const selectedFile = useSelector(selectActiveFile);
+  const preferences = useSelector((state) => state.preferences);
+  const project = useSelector((state) => state.project);
+  const isUserOwner = useSelector(getIsUserOwner);
+  const dispatch = useDispatch();
+  const { t } = useTranslation();
 
-    this.state = {
-      consoleSize: props.ide.consoleIsExpanded ? 150 : 29,
-      sidebarSize: props.ide.sidebarIsExpanded ? 160 : 20
-    };
-  }
+  const [consoleSize, setConsoleSize] = useState(150);
+  const [sidebarSize, setSidebarSize] = useState(160);
+  const [isOverlayVisible, setIsOverlayVisible] = useState(true);
 
-  componentDidMount() {
-    // If page doesn't reload after Sign In then we need
-    // to force cleared state to be cleared
-    this.props.clearPersistedState();
+  const cmRef = useRef({});
 
-    this.props.stopSketch();
-    if (this.props.params.project_id) {
-      const { project_id: id, username } = this.props.params;
-      if (id !== this.props.project.id) {
-        this.props.getProject(id, username);
-      }
-    }
+  const autosaveIntervalRef = useRef(null);
 
-    // window.onbeforeunload = this.handleUnsavedChanges;
-    window.addEventListener('beforeunload', this.handleBeforeUnload);
+  const location = useLocation();
 
-    this.autosaveInterval = null;
-  }
+  const prevFileNameRef = useRef(selectedFile.name);
+  const locationRef = useRef(location.pathname);
 
-  componentWillReceiveProps(nextProps) {
-    if (nextProps.location !== this.props.location) {
-      this.props.setPreviousPath(this.props.location.pathname);
-    }
-    if (this.props.ide.sidebarIsExpanded !== nextProps.ide.sidebarIsExpanded) {
-      this.setState({
-        sidebarSize: nextProps.ide.sidebarIsExpanded ? 160 : 20
-      });
-    }
-  }
+  const syncFileContent = () => {
+    const file = cmRef.current.getContent();
+    dispatch(updateFileContent(file.id, file.content));
+  };
+
+  useEffect(() => {
+    dispatch(clearPersistedState());
 
-  componentWillUpdate(nextProps) {
-    if (nextProps.params.project_id && !this.props.params.project_id) {
-      if (nextProps.params.project_id !== nextProps.project.id) {
-        this.props.getProject(nextProps.params.project_id);
+    dispatch(stopSketch());
+    if (props.params.project_id) {
+      const { project_id: id, username } = props.params;
+      if (id !== project.id) {
+        dispatch(getProject(id, username));
       }
     }
-  }
+  }, []);
+
+  // for setting previous location
+  useEffect(() => {
+    dispatch(setPreviousPath(locationRef.current.pathname));
+  }, [location.pathname]);
 
-  componentDidUpdate(prevProps) {
-    if (this.props.isUserOwner && this.props.project.id) {
+  // For autosave
```

Comment:
Let's define some of the complicated `boolean` logic outside of the `useEffect` so that we can minimize the dependencies that we need.

```suggestion
  const autosaveAllowed = isUserOwner && project.id && preferences.autosave;
  const shouldAutosave = autosaveAllowed && ide.unsavedChanges;

  // For autosave - send to API after 5 seconds without changes
```